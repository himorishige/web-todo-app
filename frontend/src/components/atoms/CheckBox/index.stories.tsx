/** @jsxImportSource @emotion/react */
import { ComponentStory, ComponentMeta } from '@storybook/react';
import CheckBox from '.';

export default {
  title: 'WebToDoApp/Atoms/CheckBox',
  component: CheckBox,
} as ComponentMeta<typeof CheckBox>;

const Template: ComponentStory<typeof CheckBox> = (args) => <CheckBox {...args} />;

export const Checked = Template.bind({});
Checked.storyName = 'Checked';
Checked.args = {
  disabled: false,
  isCompleted: true,
};

export const UnChecked = Template.bind({});
UnChecked.storyName = 'UnChecked';
UnChecked.args = {
  disabled: false,
  isCompleted: false,
};
