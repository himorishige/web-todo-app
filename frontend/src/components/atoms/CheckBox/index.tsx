/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { memo } from 'react';

export type Props = {
  isCompleted?: boolean;
  disabled?: boolean;
  onChange?: () => void;
};

const CheckBox: React.VFC<Props> = memo((props) => {
  const { isCompleted = false, disabled = false, onChange } = props;

  return (
    <>
      <label css={checkBoxStyle}>
        <input
          type="checkbox"
          className="input"
          disabled={disabled}
          defaultChecked={isCompleted}
          onChange={onChange}
        />
        <span className="dummy-input" data-testid="checkbutton"></span>
      </label>
    </>
  );
});

export default CheckBox;

const checkBoxStyle = css`
  width: 100%;
  display: flex;
  align-items: center;
  cursor: pointer;

  .input {
    margin: 0;
    width: 0;
    opacity: 0;
  }

  .input:hover {
    background: rgba(0, 0, 0, 0.05) !important;
  }

  .input:hover > .dummy-input {
    transform: scale(1.1);
  }

  .input:focus + .dummy-input {
    transform: scale(1.1);
  }
  .input:checked + .dummy-input {
    background: var(--primary-color);
  }
  .input:checked + .dummy-input::before {
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: 50%;
    width: 35%;
    height: 4px;
    border-radius: 2px;
    transform: translate(-6px, 5px) rotateZ(-135deg);
    transform-origin: 2px 2px;
    background: #ffffff;
  }
  .input:checked + .dummy-input::after {
    content: '';
    display: block;
    position: absolute;
    top: 50%;
    left: 50%;
    width: 70%;
    height: 4px;
    border-radius: 2px;
    transform: translate(-6px, 5px) rotateZ(-45deg);
    transform-origin: 2px 2px;
    background: #ffffff;
  }
  .dummy-input {
    position: relative;
    top: 0;
    left: 0;
    display: block;
    width: 32px;
    height: 32px;
    border: solid 2px transparent;
    background: rgb(216, 179, 132, 0.5);
    border-radius: 50%;
    transition: all 0.15s linear;
  }
`;
