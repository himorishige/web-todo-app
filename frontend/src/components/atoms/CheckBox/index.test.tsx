import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CheckBox from '.';
import { matchers } from '@emotion/jest';

// Add the custom matchers provided by '@emotion/jest'
expect.extend(matchers);

describe('Atoms/CheckBox', () => {
  test('初期状態は未チェックの状態になる', () => {
    render(<CheckBox />);
    const element = screen.getByTestId('checkbutton');
    expect(element).toHaveStyle('background: rgb(216, 179, 132, 0.5)');
    const input = screen.getByRole('checkbox');
    expect(input).toHaveProperty('defaultChecked', false);
  });

  test('isCompleted:trueの場合にチェックの状態になる', () => {
    render(<CheckBox isCompleted={true} />);
    const input = screen.getByRole('checkbox');
    expect(input).toHaveProperty('defaultChecked', true);
    const element = screen.getByRole('checkbox');
    expect(element).toHaveStyle('background: var(--primary-color)');
  });

  test('正しくレンダリングされている', () => {
    const target = renderer.create(<CheckBox />).toJSON();
    expect(target).toMatchSnapshot();
  });
});
