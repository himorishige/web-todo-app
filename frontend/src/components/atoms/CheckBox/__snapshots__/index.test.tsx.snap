// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/CheckBox 正しくレンダリングされている 1`] = `
<label
  className="css-dygvpw"
>
  <input
    className="input"
    defaultChecked={false}
    disabled={false}
    type="checkbox"
  />
  <span
    className="dummy-input"
    data-testid="checkbutton"
  />
</label>
`;
